File: BOJ/2606_바이러스.py
# 230122
# BFS, DFS

'''
# ver1. BFS

from collections import deque

n = int(input())    # node
v = int(input())    # edge
graph = [[] for i in range(n+1)]
visited = [0]*(n+1)

# print("visited = ", visited)

for i in range(v):  # create graph
    a, b = map(int, input().split())
    graph[a] += [b]
    graph[b] += [a]

visited[1] = 1
queue = deque([1])

while queue:
    c = queue.popleft()
    for nx in graph[c]:
        if visited[nx] == 0:
            queue.append(nx)
            visited[nx] = 1

print(sum(visited)-1)
'''

# ver2. DFS

n = int(input())    # node
v = int(input())    # edge
graph = [[] for i in range(n+1)]
visited = [0]*(n+1)

for i in range(v):  # create graph
    a, b = map(int, input().split())
    graph[a] += [b]
    graph[b] += [a]


def dfs(c):
    visited[c] = 1
    for nx in graph[c]:
        if visited[nx] == 0:
            dfs(nx)

dfs(1)

print(sum(visited)-1)

File: BOJ/2606_바이러스.cpp
// 230529

#include <iostream>
#include <vector>
using namespace std;

int n, v;
vector<vector<int>> graph;
vector<int> visited;

void dfs(int idx){
    visited[idx] = 1;
    for(int i=0;i<graph[idx].size();i++){
        if (visited[graph[idx][i]] == 0) dfs(graph[idx][i]);
    }
}

int main() {
    cin >> n >> v;
    
    visited = vector<int> (n+1, 0);
    graph = vector<vector<int>> (n+1);
    int s, e;
    
    for(int i = 0;i < v; i++){
        cin >> s >> e;
        graph[s].push_back(e);
        graph[e].push_back(s);
    }
    
    dfs(1);
    
    int total = 0;
    for (int i=1;i<=n;i++){
        if (visited[i] == 1) total++;
    }
    
    cout << total - 1;
    
}
